File: experiments/runs/2026-09-20_06-59-55_claim-office-java-example-mapping_exact-ptdd-v1-cc_opus-5-no-thinking/src/main/java/Policy.java
import java.util.List;
import java.util.Map;

/** An MHPCO policy: the insured items and how much of the payout cap is still available. */
final class Policy {

    private static final int CAP_FACTOR = 2;

    private final List<Map<String, Object>> insuredItems;
    private int remainingCap;

    Policy(List<Map<String, Object>> insuredItems, int insuranceSum) {
        this.insuredItems = List.copyOf(insuredItems);
        this.remainingCap = insuranceSum * CAP_FACTOR;
    }

    IncidentCoverage coverageForIncident() {
        return new IncidentCoverage(insuredItems);
    }

    int remainingCap() {
        return remainingCap;
    }

    int drawFromCap(double desiredPayout) {
        int payout = (int) Math.floor(Math.min(desiredPayout, remainingCap));
        remainingCap -= payout;
        return payout;
    }
}

File: experiments/runs/2026-09-23_05-50-16_claim-office-example-mapping_baseline-inline-tdd-v1.1-local-git-cc_opus-5-no-thinking/src/policy.ts
import { Item, insuranceSum } from './quote.js';
import { ClaimResult, Damage, InvalidDamageError, UninsuredDamageError } from './claim.js';

const DEDUCTIBLE = 100;
const CAP_FACTOR = 2;
const HIGH_ENCHANTMENT_CLAIM_THRESHOLD = 8;
const HIGH_ENCHANTMENT_REIMBURSEMENT = 0.5;

/**
 * A policy created by a quote step. It knows the insured items and tracks how
 * much of the payout cap successive claims have already consumed.
 */
export class Policy {
  readonly insuranceSum: number;
  private capLeft: number;

  constructor(private readonly items: Item[]) {
    this.insuranceSum = insuranceSum(items);
    this.capLeft = this.insuranceSum * CAP_FACTOR;
  }

  get remainingCap(): number {
    return this.capLeft;
  }

  claim(damages: Damage[]): ClaimResult {
    const damagedItems = this.matchDamagesToItems(damages);

    let desired = 0;
    for (const { item, amount } of damagedItems) {
      desired += Math.max(0, this.reimbursement(item, amount) - DEDUCTIBLE);
    }

    const payout = Math.floor(Math.min(desired, this.capLeft));
    this.capLeft -= payout;
    return { payout, remainingCap: this.capLeft };
  }

  /**
   * Each damage entry is a separate damage against one insured item, so a
   * policy covering two swords absorbs two sword damages but no more.
   */
  private matchDamagesToItems(damages: Damage[]): { item: Item; amount: number }[] {
    const available = [...this.items];

    return damages.map(({ itemType, amount }) => {
      if (amount < 0) throw new InvalidDamageError(`negative damage amount: ${amount}`);

      const index = available.findIndex((item) => item.type === itemType);
      if (index === -1) {
        throw new UninsuredDamageError(`damage to an item not covered by the policy: ${itemType}`);
      }

      const [item] = available.splice(index, 1);
      return { item, amount };
    });
  }

  /**
   * Damage is reimbursed in full - as the dragon-material clause spells out
   * and as the standard case does anyway - except for highly enchanted items,
   * whose 50 % clause wins even when the item is also dragon material.
   */
  private reimbursement(item: Item, amount: number): number {
    return (item.enchantment ?? 0) >= HIGH_ENCHANTMENT_CLAIM_THRESHOLD
      ? amount * HIGH_ENCHANTMENT_REIMBURSEMENT
      : amount;
  }
}
